### DISCUSSION TITLE
Default Value for Language Feature

### DISCUSSION THREAD

**Discussion by ballcoach12** [eclipse-langium/langium#352] [Q&A] [answered]
I need a way to set default values for language features that are optional in a grammar. In Xtext, you could use a PostProcessor to do this. Is there a way to do this in Langium? I could imagine that it might be accomplished in the grammar definition itself, maybe like this:

```
Greeting:
'Hello, " (name=STRING {'Charlie'})?;
```

**Comment by msujew** [comment, marked as answer]
Hi @ballcoach12,

Interesting proposal, currently there's no built-in way to accomplish this.

However, before processing your AST content you could simply run:
```ts
streamAllContents(rootElement).forEach(e => {
  if (isGreeting(e.node) && e.node.name === undefined) {
    e.node.name = 'Charlie'; 
  }
})
```

This should accomplish the same thing for now.

**Reply by ballcoach12** [reply]
To implement your recommendation, my assumption is that I'll need to do my post processing immediately after parsing. In order to do that, I expect that I'm going to need to supply my own service implementation that extends a default service class and inject it in the module. Would AstNodeDescriptionProvider be the class that I supply an alternative implementation for? 

Lots of assumptions on my part here :).

**Reply by spoenemann** [reply]
I would do that by overriding `documents.LangiumDocumentFactory` (v0.3: `workspace.LangiumDocumentFactory` in the shared services) with  a subclass of `DefaultLangiumDocumentFactory`. There you can do any post-processing you want on the AST. But be careful not to follow any cross-references while doing so, since that is only possible after a later build stage of the document.

Taking a step back: What is your motivation for adding default values? Is it for convenience when using the resulting AST?

**Reply by ballcoach12** [reply]
My grammar specifies several properties that are optional. The goal is to keep the concrete syntax brief while also keeping the number of grammar rules small.

In this particular case, our code generation routines could pick up the AST and fill in the default values in the generated code, but we also have other languages that depend on there being a property value that is not undefined in order for validation to work properly.

**Reply by ballcoach12** [reply]
@spoenemann, I am finally getting back to working on this. Here is what I came up with, which I think aligns with what you were recommending:

```
export class FeederDocumentFactory extends DefaultLangiumDocumentFactory {

    constructor(services: LangiumSharedServices) {
        super(services);
    }

    override fromTextDocument<T extends AstNode = AstNode>(textDocument: TextDocument): LangiumDocument<T> {
        var returnValue = super.fromTextDocument<T>(textDocument);
        let rootElement = returnValue.parseResult.value;
        if (isFeeder(rootElement)) {
            if(rootElement.uniqueId === undefined){
                rootElement.uniqueId = generateUniqueId();
            }
            if (rootElement.eventDetSecs === undefined) {
                rootElement.eventDetSecs = 3;
            }
            if (rootElement.isLossSourceDetEna === undefined) {
                rootElement.isLossSourceDetEna = false;
            }
            if (rootElement.isMiscoordDetEna === undefined) {
                rootElement.isMiscoordDetEna = false;
            }
            if (rootElement.isOpenPhaseDetEnab === undefined) {
                rootElement.isOpenPhaseDetEnab = false;
            }
            if (rootElement.nomVolt === undefined) {
                rootElement.nomVolt = 12470;
            }
            if (rootElement.openPhaseDetSecs === undefined) {
                rootElement.openPhaseDetSecs = 3;
            }
            if (rootElement.rtrnToNormalMode === undefined) {
                rootElement.rtrnToNormalMode = "Disabled";
            }
            if (rootElement.shouldIsoOpenPhase === undefined) {
                rootElement.shouldIsoOpenPhase = false;
            }
           
        }
        streamAllContents(rootElement).forEach(e => {
            if (isEdgeStmt(e)) {
                if (e.uniqueId === undefined) {
                    e.uniqueId = generateUniqueId();
                }
            }
            if (isEdgeRhs(e)) {
                if (e.uniqueId === undefined) {
                    e.uniqueId = generateUniqueId();
                }
                if (e.loadCapPercent === undefined) {
                    e.loadCapPercent = 100;
                }
                if (e.tempCapPercent === undefined) {
                    e.tempCapPercent = 100;
                }
            }
            if (isDevice(e)) {
                if (e.uniqueId === undefined) {
                    e.uniqueId = generateUniqueId();
                }
                if (e.isBrkr || e.isRecloser || e.isSwitch) {
                    if (e.loadCap === undefined) {
                        e.loadCap = 12134567;
                    }
                    if (e.tempCap === undefined) {
                        e.tempCap = 100;
                    }
                    if (e.normalStatus === undefined) {
                        e.normalStatus = "Closed";
                    }
                    if (e.leftPTPosition === null) {
                        e.leftPTPosition = 'None';
                    }
                    if (e.rightPTPosition === null) {
                        e.rightPTPosition = 'None';
                    }
                }
                if (e.isNeighboringFeeder) {
                    if (e.loadCap === undefined) {
                        e.loadCap = 100;
                    }
                    if (e.tempCap === undefined) {
                        e.tempCap = 100;
                    }
                }
            }
        })
        return returnValue;
    }
}
```

and here is  how I inject it:

```
export const FlisrFeederDslModule: Module<FlisrFeederDslServices, PartialLangiumServices & FlisrFeederDslAddedServices> = {
    validation: {
        ValidationRegistry: (services) => new FlisrFeederDslValidationRegistry(services),
        FlisrFeederDslValidator: () => new FlisrFeederDslValidator()
    }
    ,
    shared: {
        workspace: {
            LangiumDocumentFactory: (services) => new FeederDocumentFactory(services.shared)
        }
    }
};
```
The problem is that in the `create()` function in DefaultLangiumDocumentFactory cannot find a "serviceRegistry".  Maybe I'm not injecting things correctly? 

I have a unit test that calls  the  following function  in my module:

```
export function createFlisrFeederDslServices(context?: DefaultSharedModuleContext): {
    shared: LangiumSharedServices,
    FlisrFeederDsl: FlisrFeederDslServices
} {
    const shared = inject(
        createDefaultSharedModule(context),
        FlisrFeederDslGeneratedSharedModule
    );
    const FlisrFeederDsl = inject(
        createDefaultModule({ shared }),
        FlisrFeederDslGeneratedModule,
        FlisrFeederDslModule
    );
    shared.ServiceRegistry.register(FlisrFeederDsl);

    
    return { shared, FlisrFeederDsl };
}
```
Not sure what I am doing wrong  in this case.

**Reply by spoenemann** [reply]
The shared services are created in a separate call to `inject`. You need to create a new module:
```
export const FlisrFeederDslSharedModule: Module<LangiumSharedServices, PartialLangiumSharedServices> = {
    workspace: {
        LangiumDocumentFactory: (services) => new FeederDocumentFactory(services)
    }
};
```
and add this to the respective `inject` call.

**Reply by ballcoach12** [reply]
Perfect! Thank you! That was the missing piece!

**Reply by Yokozuna59** [reply]
Hi there, I came across the same question, but the current approach would still make the attribute as optional. So, from a design perspective, would you recommend extending the interface to make the attribute as defined? Using `result.value?.attribute` isn't the best choice.

**Reply by msujew** [reply]
Rethinking this, it might make sense to add default values to declared interfaces. I'm imagining something like this - which should also keep the performance overhead at a minimum:

```typescript
interface Item {
  value: string;
  version: string = "2.0.0"
}
```

I've created an issue for this over at https://github.com/eclipse-langium/langium/issues/1154.
